[DESIGN REVIEW][WIP][FEATURE]: Typed validation operator results

roblim opened the pull request:
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [ENHANCEMENT], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- New `ValidationOperatorResult` for validation operator results

After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
- https://github.com/great-expectations/great_expectations/pull/1311/

Thank you for submitting!

roblim committed:
Add ValidationOperatorResult type

roblim committed:
Nix unreachable code

roblim committed:
Merge branch 'develop' into feature/typed-validation-operator-results

roblim committed:
Use ValidationOperatorResult

roblim committed:
Update ValidationOperatorResult

roblim committed:
Nix old code

roblim committed:
Add properties and getters

- self._data_asset_names
- self._actions_results_by_validation_result_identifier
- self._validation_results_by_expectation_suite_name
- self._validation_results_by_data_asset_name

roblim committed:
Fix typos

roblim committed:
Add validation operator config to typed result; add name attribute

roblim committed:
Merge branch 'develop' into feature/typed-validation-operator-results

roblim commented on great_expectations/validation_operators/validation_operators.py:
```diff
@@ -27,6 +31,13 @@ class ValidationOperator(object):
+        self._validation_operator_config = None
+
+    @property
+    def validation_operator_config(self):
```
Suggestion to rename this to `checkpoint_config`

ayirplm replied:
I think Eugene was going to go over semantic uses of validation_operator vs checkpoint

roblim committed:
Update WarningAndFailureExpectationSuitesValidationOperator to return typed result

roblim committed:
Add batch_identifiers property; add optional success param

roblim committed:
Nix placeholder docstring

roblim committed:
Add docstrings

roblim committed:
Add statistics

roblim committed:
Merge branch 'develop' into feature/typed-validation-operator-results

roblim commented on great_expectations/validation_operators/types/validation_operator_result.py:
```diff
@@ -0,0 +1,258 @@
+    "warning" or "failure" level.
+
+    e.g.
+    {
```
This is basic contract for what this type expects to get from a `ValidationOperator` run. Some notes/?'s:
- top-level keys are just the `ValidationResultIdentifiers` from the run
- corresponding value is a dict with at least a `validation_result` key with the actual `ExpectationSuiteValidationResult` and an `actions_results` key (this is dict of all action outputs from validation)
- all other properties are derived from previous, lazily
- new validation operators may add additional keys here as needed, in which case their values would only be accessible from the `run_results` key

- right now, the only two validation operators are `action_list` based, hence the inclusion of the `actions_results` key in the basic contract. I can imagine this not being true in the future though. Should we keep this in the basic contract? If not, it would just mean the `actions_results_by_validation_result_identifier` property goes away and is only accessible from the `run_results` property

ayirplm replied:
Basic contract too. There maybe folks who are using actions_results_by_validation_result_identifier currently.

roblim commented on great_expectations/validation_operators/validation_operators.py:
```diff
@@ -22,11 +26,47 @@ class ValidationOperator(object):
+            },
+        }
+
+        {
```
Is this the right approach for validation operator (or checkpoint) config?

Aylr replied:
I'm not sure - personally, this would be easier to discuss synchronously.

alexsherstinsky commented on great_expectations/validation_operators/types/validation_operator_result.py:
```diff
@@ -0,0 +1,258 @@
+                    ]
+                )
+            )
+        return self.batch_identifiers
```
Are we returning self.batch_identifiers or self._batch_identifiers?..

roblim commented on great_expectations/validation_operators/types/validation_operator_result.py:
```diff
@@ -0,0 +1,258 @@
+            ) * 100
+
+            self._statistics = {
+                "data_asset_count": data_asset_count,
```
What else would be useful here?

ayirplm replied:
Trend lines across batches

Aylr replied:
It might be nice to expose one level deeper of statistics such as the statistics found in each individual validation result. So without much work a user could generate a simple report that shows the number of expectations in each validation that passed.

Just spit-balling here, I could imagine something like:
```python
validation_result.statistics.details_by_suite
{
    "foo.warning": [ValidationResult1_stats],
    "foo.critical": [ValidationResult1_stats, ValidationResult2_stats],
 }
```

The intent is to make it easier to get detailed stats from such a high level aggregate without forcing the user to de-nest them themselves.

alexsherstinsky reviewed (commented):
I found one potential issue.  Can review again once all comments have been reviewed.  Thanks!

roblim commented on great_expectations/validation_operators/types/validation_operator_result.py:
```diff
@@ -0,0 +1,258 @@
+                    "expectation_suite_name"
+                ]
+                if batch_id not in assets_validated_by_batch_id:
+                    assets_validated_by_batch_id[batch_id] = {
```
@Aylr - would this have helped in generating that client request (copied below)? Or perhaps one of: self._validation_results_by_expectation_suite_name, self._validation_results_by_data_asset_name

```
Quality Expectations, Profiles and Operational Check Details: https://s3.console.aws.amazon.com/s3/buckets/ge-data-documents
Please review the expectation failure summaries below:
- operational_check_1: Failed! 90.0 % passed (9 of 10 expectations were met)
- operational_check_2: Failed! 90.0 % passed (9 of 10 expectations were met)
- operational_check_3: Failed! 90.0 % passed (9 of 10 expectations were met)
- operational_check_4: Passed! 100.0 % passed (10 of 10 expectations were met)
```

ayirplm replied:
All as above

Aylr replied:
Rob this would have helped tremendously! I like that there are multiple ways to get at this - via the suite names or via the data asset names. Nice!

roblim reviewed (commented):
> I found one potential issue. Can review again once all comments have been reviewed. Thanks!

Good catch!

roblim committed:
Fix typo

roblim committed:
Merge branch 'feature/typed-validation-operator-results' of github.com:great-expectations/great_expectations into feature/typed-validation-operator-results

ayirplm reviewed (commented):
Inline comments

Aylr commented on great_expectations/validation_operators/types/validation_operator_result.py:
```diff
@@ -0,0 +1,258 @@
+        evaluation_parameters: dict = None,
+        success: bool = None,
+    ) -> None:
+        self._run_id = run_id
```
I don't know if it's possible, but it might be cool to have a property that contains urls links to built data docs sites for this run. That would be helpful for building richer notifications.

ayirplm replied:
get_docs_sites_urls is valid for this too

Aylr reviewed (commented):
Design review comments and questions. I did not look at much implementation - mostly the names of properties, methods and your excellent questions @roblim .

jcampbell reviewed (commented):
I'm really excited about providing more useful structure for users here.

I have a question: given that this class is really exposing lots of pseudo-analytic pivots over the validation results, I wonder whether it makes more sense to redefine what are currently exposed as properties as explicit queries over the validation operator result. -- "list" and "get" operations, for example. Thoughts?

roblim committed:
Merge branch 'develop' into feature/typed-validation-operator-results

roblim committed:
Nix actions_results from run_results basic contract

roblim committed:
Convert properties to list_ pattern; nix self._actions_results_by_validation_result_identifier

Reduce properties to:
self._run_id
self._run_results
self._evaluation_parameters
self._validation_operator_config
self._success

roblim committed:
Add validation statistics

roblim committed:
Prevent division by zero

roblim committed:
Add ValidationOperatorResultSchema and to_json_dict method

roblim commented:
Closing in favor of https://github.com/great-expectations/great_expectations/pull/1472
